File: ios-native/Jottery/Jottery/Models/NoteRecord.swift
import Foundation
import GRDB

/// On-disk note record with encrypted fields. Maps directly to the `notes` SQLite table.
/// Decryption happens in `NoteRepository`, not here.
struct NoteRecord: Codable, FetchableRecord, PersistableRecord, Identifiable {
    static let databaseTableName = "notes"

    // MARK: - Columns

    var id: String
    var createdAt: String          // ISO 8601
    var modifiedAt: String         // ISO 8601
    var syncedAt: String?
    var content: String            // Encrypted JSON: {"ciphertext":"...","iv":"..."}
    var tags: String               // Encrypted JSON: {"ciphertext":"...","iv":"..."}
    var attachments: String        // JSON array of attachment references
    var pinned: Bool
    var archived: Bool
    var archivedAt: String?
    var locked: Bool
    var lockedAt: String?
    var deleted: Bool
    var deletedAt: String?
    var version: Int
    var wordWrap: Bool
    var syntaxLanguage: String
    var showPreview: Bool
    var color: String?
    var contentHash: String?
    var parentHash: String?
    var hashChain: String?         // JSON array of hash strings
    var needsSync: Bool

    // MARK: - Column Mapping

    enum Columns: String, ColumnExpression {
        case id
        case createdAt = "created_at"
        case modifiedAt = "modified_at"
        case syncedAt = "synced_at"
        case content
        case tags
        case attachments
        case pinned
        case archived
        case archivedAt = "archived_at"
        case locked
        case lockedAt = "locked_at"
        case deleted
        case deletedAt = "deleted_at"
        case version
        case wordWrap = "word_wrap"
        case syntaxLanguage = "syntax_language"
        case showPreview = "show_preview"
        case color
        case contentHash = "content_hash"
        case parentHash = "parent_hash"
        case hashChain = "hash_chain"
        case needsSync = "needs_sync"
    }

    // MARK: - Coding Keys (snake_case ↔ camelCase)

    enum CodingKeys: String, CodingKey {
        case id
        case createdAt = "created_at"
        case modifiedAt = "modified_at"
        case syncedAt = "synced_at"
        case content
        case tags
        case attachments
        case pinned
        case archived
        case archivedAt = "archived_at"
        case locked
        case lockedAt = "locked_at"
        case deleted
        case deletedAt = "deleted_at"
        case version
        case wordWrap = "word_wrap"
        case syntaxLanguage = "syntax_language"
        case showPreview = "show_preview"
        case color
        case contentHash = "content_hash"
        case parentHash = "parent_hash"
        case hashChain = "hash_chain"
        case needsSync = "needs_sync"
    }

    // MARK: - Factory

    /// Create a new empty note record with encrypted empty content.
    static func new(encryptedContent: String, encryptedTags: String) -> NoteRecord {
        let now = ISO8601DateFormatter.jottery.string(from: Date())
        return NoteRecord(
            id: CryptoService.generateUUID(),
            createdAt: now,
            modifiedAt: now,
            syncedAt: nil,
            content: encryptedContent,
            tags: encryptedTags,
            attachments: "[]",
            pinned: false,
            archived: false,
            archivedAt: nil,
            locked: false,
            lockedAt: nil,
            deleted: false,
            deletedAt: nil,
            version: 0,
            wordWrap: true,
            syntaxLanguage: "markdown",
            showPreview: false,
            color: nil,
            contentHash: nil,
            parentHash: nil,
            hashChain: nil,
            needsSync: true
        )
    }
}

File: android-native/app/src/main/java/com/jottery/android/model/NoteRecord.kt
package com.jottery.android.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.jottery.android.util.DateUtils

/**
 * On-disk note record with encrypted fields. Maps directly to the `notes` SQLite table.
 * Decryption happens in NoteRepository, not here.
 */
@Entity(tableName = "notes")
data class NoteRecord(
    @PrimaryKey
    val id: String,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "modified_at")
    var modifiedAt: String,

    @ColumnInfo(name = "synced_at")
    var syncedAt: String? = null,

    // Encrypted JSON: {"ciphertext":"...","iv":"..."}
    var content: String,

    // Encrypted JSON: {"ciphertext":"...","iv":"..."}
    var tags: String,

    // JSON array of attachment references
    var attachments: String = "[]",

    var pinned: Boolean = false,

    var archived: Boolean = false,

    @ColumnInfo(name = "archived_at")
    var archivedAt: String? = null,

    var locked: Boolean = false,

    @ColumnInfo(name = "locked_at")
    var lockedAt: String? = null,

    var deleted: Boolean = false,

    @ColumnInfo(name = "deleted_at")
    var deletedAt: String? = null,

    var version: Int = 1,

    @ColumnInfo(name = "word_wrap")
    var wordWrap: Boolean = true,

    @ColumnInfo(name = "syntax_language")
    var syntaxLanguage: String = "markdown",

    @ColumnInfo(name = "show_preview")
    var showPreview: Boolean = false,

    var color: String? = null,

    @ColumnInfo(name = "content_hash")
    var contentHash: String? = null,

    @ColumnInfo(name = "parent_hash")
    var parentHash: String? = null,

    @ColumnInfo(name = "hash_chain")
    var hashChain: String? = null,

    @ColumnInfo(name = "needs_sync")
    var needsSync: Boolean = true,

    @ColumnInfo(name = "sync_hash")
    var syncHash: String? = null,
) {
    companion object {
        fun new(encryptedContent: String, encryptedTags: String): NoteRecord {
            val now = DateUtils.nowISO8601()
            return NoteRecord(
                id = java.util.UUID.randomUUID().toString().lowercase(),
                createdAt = now,
                modifiedAt = now,
                content = encryptedContent,
                tags = encryptedTags,
                attachments = "[]",
                pinned = false,
                archived = false,
                locked = false,
                deleted = false,
                version = 0,
                wordWrap = true,
                syntaxLanguage = "markdown",
                showPreview = false,
                needsSync = true,
            )
        }
    }
}
